### ISSUE TITLE
Regression ServerRequestContext::currentRequest not working in HTTP Filters since 2.5.0

### ISSUE THREAD

**Issue by sdelamo** [micronaut-projects/micronaut-core#5400]
`ServerRequestContext.currentRequest();`  is not resolving the request since 2.5.0 in a Http Server Filter. 

Given the following controllers: 

```java
package example.micronaut;

import io.micronaut.http.MediaType;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.Produces;

@Controller("/es")
public class EsController {

    @Produces(MediaType.TEXT_HTML)
    @Get
    public String index() {
        return "es";
    }
}
```

```java
package example.micronaut;

import io.micronaut.http.MediaType;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.Produces;

@Controller("/en")
public class EnController {

    @Produces(MediaType.TEXT_HTML)
    @Get
    public String index() {
        return "en";
    }
}
```

```java
package example.micronaut;

import io.micronaut.core.async.publisher.Publishers;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.MutableHttpResponse;
import io.micronaut.http.annotation.Filter;
import io.micronaut.http.context.ServerRequestContext;
import io.micronaut.http.filter.OncePerRequestHttpServerFilter;
import io.micronaut.http.filter.ServerFilterChain;
import org.reactivestreams.Publisher;
import java.net.URI;
import java.util.Optional;

@Filter(Filter.MATCH_ALL_PATTERN)
public class LangFilter extends OncePerRequestHttpServerFilter {
    @Override
    protected Publisher<MutableHttpResponse<?>> doFilterOnce(HttpRequest<?> request, ServerFilterChain chain) {
        if (request.getUri().toString().equals("/es")) {
            return chain.proceed(request);
        }
        Optional<HttpRequest<Object>> optionalHttpRequest = ServerRequestContext.currentRequest();
        if (optionalHttpRequest.isPresent()) {
            String accepts = optionalHttpRequest.get().getHeaders().get("Accept-Language");
            if (accepts != null && accepts.equals("es")) {
                return Publishers.just(HttpResponse.seeOther(URI.create("/es")));
            }
        }
        return chain.proceed(request);
    }
}
```

With 2.4.4, I am able to get hold of the current request with `ServerRequestContext`

```
curl -iL -H "Accept-Language: es" localhost:8080/en
HTTP/1.1 303 See Other
Location: /es
Date: Sat, 8 May 2021 02:47:15 GMT
connection: keep-alive
transfer-encoding: chunked

HTTP/1.1 200 OK
Date: Sat, 8 May 2021 02:47:15 GMT
content-type: text/html
content-length: 2
connection: keep-alive

es
```

with 2.5.0 or 2.5.1, it always returns optional: 

```
curl -iL -H "Accept-Language: es" localhost:8080/en
HTTP/1.1 200 OK
Date: Sat, 8 May 2021 02:50:05 GMT
content-type: text/html
content-length: 2
connection: keep-alive

en
```

**Comment by graemerocher** [comment]
Agree with the regression but why would you need to if the request is already an argument. 🤔

**Comment by graemerocher** [comment]
I guess yeah if there http clients used may need to be picked up downstream. Definitely need to fix this.

**Comment by andruskutt** [comment]
This might have caused by filter ordering issue: On my custom filter implementing HttpServerFilter method Ordered.getOrder() is not called anymore on micronaut 2.5.X

**Comment by sdelamo** [comment]
> Agree with the regression but why would you need to if the request is already an argument. 🤔

I found it when using a `TenantResolver::resolveIdentifier()`within a Filter. I tried to create the simplest reproducible example.

**Comment by andruskutt** [comment]
Filter ordering fix seems to be incomplete. I have DefaultRouter.preconditionFilterRoutes empty, only two filters in alwaysMatchesFilterRoutes / alwaysMatchesHttpFilters . And they are not sorted by order: my filter, which depends on SecurityFilter ( ServerFilterPhase.SECURITY.after() ) , is first in list.

**Comment by jameskleeh** [comment]
@andruskutt You would have to upload an example. We have tests that verify the ordering works

**Comment by andruskutt** [comment]
[issue-5400.zip](https://github.com/micronaut-projects/micronaut-core/files/6461840/issue-5400.zip)
Just run gradlew test.

**Comment by jameskleeh** [comment]
@andruskutt Thanks for the example. It appears the issue would not happen if any filter matched a subset of requests. Will be fixed in 2.5.3 coming soon
